explorer/testdata/struct/fail_assign_different_types.carbon

```diff
@@ -0,0 +1,16 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// RUN: %{not} %{explorer-run}
+// RUN: %{not} %{explorer-run-trace}
+
+package ExplorerTest api;
+
+fn Main() -> i32 {
+  var p: auto = {.x = 0, .y = 0};
+  // CHECK:STDERR: COMPILATION ERROR: {{.*}}/explorer/testdata/struct/fail_assign_different_types.carbon:[[@LINE+1]]: type error in assignment: '{.z: i32, .y: i32}' is not implicitly convertible to '{.x: i32, .y: i32}'
+  p = {.z = 0, .y = 0};
```

Comment:
So, I think this gets to the crux of the test: won't this fail because of the logic on 368, because a destination field won't be found for `z`?

How about this: do `{.y = 0,}`, but remove the `.size()` check on 356. Add a TODO to your new code that it should skip defaulted fields once they're added.

I say this because the `.size()` check will need to be removed anyways once defaulted fields are supported, and your new code should provide equivalent logic.